### PR TITLE
dev-cmd/pr-publish: pass `--autosquash` by default

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

It's too easy to forget to, and the consequences of forgetting to pass
`--autosquash` are worse (i.e. almost always permanent) than forgetting
to pass `--no-autosquash` (i.e. just do `pr-publish --no-autosquash`
when your command fails).

See the discussion at Homebrew/homebrew-core#80717.

### CODE REVIEW

**Review by dawidd6 — APPROVED** [review]
Need to merge this PR and quickly make changes in homebrew-core

**Comment by carlocab** [comment]
We should have the Homebrew/core PR ready to go for when this is merged. Do you mind taking care of that?

**Comment by dawidd6** [comment]
Yep. One thing: we still have pr-automerge with this option too.

**Comment by carlocab** [comment]
Hmm, yes, I think we want to make `--autosquash` the default for `pr-automerge` too. I don't think we want that for `pr-pull`; one common use case for `pr-pull` is to have finer control than `pr-publish` over how things get squashed. You also get the chance to examine what you're pushing to master with `pr-pull` unlike with `pr-publish`.

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (dawidd6) [inline Library/Homebrew/dev-cmd/pr-automerge.rb:33]:
```suggestion
      switch "--no-autosquash",
```

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (dawidd6) [inline Library/Homebrew/dev-cmd/pr-automerge.rb:46]:
```suggestion
    odeprecated "`brew pr-automerge --autosquash`" if args.autosquash?
```

**Comment by dawidd6** [comment]
But we would still have a finer control. Just with --no-autosquash option. Autosquash will fail anyway if there are multiple commits modifying multiple formulae and I believe that would be the root factor of one running it locally and trying to fixup some commits manually.

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (carlocab) [reply Library/Homebrew/dev-cmd/pr-automerge.rb]: Oops. Thanks 😄

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/pr-automerge.rb:34]:
```suggestion
             description: "Instruct `brew pr-publish` to skip automatically reformatting and rewording commits "\
```

**Comment by carlocab** [comment]
I dunno, I just think I almost always want `--no-autosquash` when doing `pr-pull`, but almost always want `--autosquash` when doing `pr-publish` or `pr-automerge`.

That said, I already have aliases that include `--autosquash` for me when doing `pr-publish`, so I guess I can do the same with `pr-pull`, so if you feel strongly about making the same changes to `pr-pull` then I don't mind.

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me, good idea!

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/pr-automerge.rb:79]: Shouldn't this be

```suggestion
    publish_args << "--no-autosquash" if args.no_autosquash?
```

Comment (carlocab) [reply Library/Homebrew/dev-cmd/pr-automerge.rb:79]: 🤦 yes, that's correct. I really should be paying more attention here. Sorry 😅

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/pr-automerge.rb:46]: Not a huge deal but I wonder if it would be helpful to show that the replacement is to just run without the arg:

```suggestion
    odeprecated "`brew pr-automerge --autosquash`", "`brew pr-automerge`" if args.autosquash?
```

That also might be too long and complicated and since this a developer command its probably not a big deal to just expect the user to look at the help text

**Review by Rylan12 — COMMENTED** [review]
I wonder if this is still a breaking change that we need to worry about. If you _were_ running `brew pr-{automerge,publish} --autosquash` it will work fine.

However, if you _were_ running `brew pr-{automerge,publish}` without `--automerge`, you'll now find `--automerge` being inferred unexpectedly. That might be okay, but you won't see any warning indicating the change. Think that's something to be concerned about?

File: Library/Homebrew/dev-cmd/pr-automerge.rb
Comment (carlocab) [reply Library/Homebrew/dev-cmd/pr-automerge.rb:46]: Yea, I don't think it hurts to show the replacement. Let's show that here and in `pr-publish`.

File: Library/Homebrew/dev-cmd/pr-publish.rb
Comment (carlocab) [inline Library/Homebrew/dev-cmd/pr-publish.rb:46]:
```suggestion
    odeprecated "`brew pr-publish --autosquash`", "`brew pr-publish`" if args.autosquash?
```

**Comment by carlocab** [comment]
> I wonder if this is still a breaking change that we need to worry about.

I don't know; I'm not too worried about it. Our `tap-new` templates only make use of `pr-pull`, which I haven't touched (though I think @dawidd6 has a different opinion here), so most third-party taps should be unaffected by these changes. The ones that will be are probably managed by users who were adventurous enough to go digging in our codebase (or were mindlessly copying Homebrew/core workflows...), so I think they'll manage. (Except for the mindless copiers. But I have to say I'm not too sympathetic here.)

**Review by Rylan12 — APPROVED** [review]

**Comment by Rylan12** [comment]
Okay. Last thing: we usually wait to add the `odeprecated` until the next major/minor version is going to be released. Should we do that here, too? If so, that would probably mean waiting on this PR, though

**Comment by carlocab** [comment]
No opinion here. I don't think this is urgent, so we can wait till the next release tag. I'm happy to defer to your or @MikeMcQuaid's opinion on this one, though.

**Comment by MikeMcQuaid** [comment]
Given this is "developer command that I'd be surprised if it's even used by Not Us": I think it's fine to merge as-is, personally.

**Comment by carlocab** [comment]
Ok, great. Merging this and Homebrew/homebrew-core#81464 then.

**Pull request merged**
